ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -725,50 +735,63 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr punish = do
       -> ChainSwitchType
       -> m (Point blk)
     switchTo vChainDiff varTentativeHeader chainSwitchType = do
-        traceWith addBlockTracer $
-            ChangingSelection
-          $ castPoint
-          $ AF.headPoint
-          $ getSuffix
+        traceWith addBlockTracer
+          . ChangingSelection
+          . castPoint
+          . AF.headPoint
+          . getSuffix
           $ getChainDiff vChainDiff
-        (curChain, newChain, events, prevTentativeHeader) <- atomically $ do
-          curChain  <- readTVar         cdbChain -- Not Query.getCurrentChain!
-          curLedger <- LgrDB.getCurrent cdbLgrDB
-          case Diff.apply curChain chainDiff of
-            -- Impossible, as described in the docstring
-            Nothing       ->
-              error "chainDiff doesn't fit onto current chain"
-            Just newChain -> do
-              writeTVar cdbChain newChain
-              LgrDB.setCurrent cdbLgrDB newLedger
-
-              -- Inspect the new ledger for potential problems
-              let events :: [LedgerEvent blk]
-                  events = inspectLedger
-                             cdbTopLevelConfig
-                             (ledgerState $ LgrDB.current curLedger)
-                             (ledgerState $ LgrDB.current newLedger)
-
-              -- Clear the tentative header
-              prevTentativeHeader <- swapTVar varTentativeHeader SNothing
-
-              case chainSwitchType of
-                -- When adding blocks, the intersection point of the old and new
-                -- tentative/selected chain is not receding, in which case
-                -- `fhSwitchFork` is unnecessary. In the case of pipelining a
-                -- block, it would even result in rolling back by one block and
-                -- rolling forward again.
-                AddingBlocks      -> pure ()
-                SwitchingToAFork -> do
-                  -- Update the followers
-                  --
-                  -- 'Follower.switchFork' needs to know the intersection point
-                  -- (@ipoint@) between the old and the current chain.
-                  let ipoint = castPoint $ Diff.getAnchorPoint chainDiff
-                  followerHandles <- Map.elems <$> readTVar cdbFollowers
-                  forM_ followerHandles $ switchFollowerToFork curChain newChain ipoint
-
-              return (curChain, newChain, events, prevTentativeHeader)
+        let lock =
```

Comment:
I'm surprised that:

- We have this, seemingly, huge lock.
- The lock depends on whether we should flush. 

Could you help me understand what's happening here? :pray: